File: src/main/swift/g0001_0100/s0024_swap_nodes_in_pairs/Solution.swift
// #Medium #Top_100_Liked_Questions #Linked_List #Recursion #Data_Structure_II_Day_12_Linked_List
// #Udemy_Linked_List #Big_O_Time_O(n)_Space_O(1)
// #2024_06_19_Time_0_ms_(100.00%)_Space_15.8_MB_(10.29%)

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init() { self.val = 0; self.next = nil; }
 *     public init(_ val: Int) { self.val = val; self.next = nil; }
 *     public init(_ val: Int, _ next: ListNode?) { self.val = val; self.next = next; }
 * }
 */
public class Solution {
    public func swapPairs(_ head: ListNode?) -> ListNode? {
        if head == nil {
            return nil
        }
        let len = getLength(head)
        return reverse(head, len)
    }

    private func getLength(_ curr: ListNode?) -> Int {
        var cnt = 0
        var curr = curr
        while curr != nil {
            cnt += 1
            curr = curr?.next
        }
        return cnt
    }

    private func reverse(_ head: ListNode?, _ len: Int) -> ListNode? {
        if len < 2 {
            return head
        }
        var curr = head
        var prev: ListNode? = nil
        var next: ListNode?
        for _ in 0..<2 {
            next = curr?.next
            curr?.next = prev
            prev = curr
            curr = next
        }
        head?.next = reverse(curr, len - 2)
        return prev
    }
}

File: src/main/dart/g0001_0100/s0024_swap_nodes_in_pairs/Solution.dart
// #Medium #Top_100_Liked_Questions #Linked_List #Recursion #Data_Structure_II_Day_12_Linked_List
// #Udemy_Linked_List #Big_O_Time_O(n)_Space_O(1)
// #2024_10_01_Time_313_ms_(89.36%)_Space_148.6_MB_(78.72%)

/**
 * Definition for singly-linked list.
 * class ListNode {
 *   int val;
 *   ListNode? next;
 *   ListNode([this.val = 0, this.next]);
 * }
 */
class Solution {
  ListNode? swapPairs(ListNode? head) {
    if (head == null) {
      return null;
    }
    int len = getLength(head);
    return reverse(head, len);
  }

  // Function to get the length of the linked list
  int getLength(ListNode? curr) {
    int cnt = 0;
    while (curr != null) {
      cnt++;
      curr = curr.next;
    }
    return cnt;
  }

  // Recursive function to reverse pairs of nodes
  ListNode? reverse(ListNode? head, int len) {
    // Base case: if less than 2 nodes, no need to swap
    if (len < 2) {
      return head;
    }
    ListNode? curr = head;
    ListNode? prev;
    ListNode? next;
    prev = null;

    // Reverse the first two nodes
    for (int i = 0; i < 2; i++) {
      next = curr!.next;
      curr.next = prev;
      prev = curr;
      curr = next;
    }

    // Recursively reverse the rest of the list
    head!.next = reverse(curr, len - 2);
    return prev;
  }
}
